server/routes/api/protein.py

```diff
@@ -25,10 +30,293 @@
+    '''
+    return dcid(reverse_interaction_id(id(interaction_dcid)))
+
+
+def filter_interaction_dcids(interaction_dcids):
+    '''
+    Given list of interaction DCIDs, remove self-interactions and deduplicate (keep one copy of each set of duplicates).
+    Self-interactions are interactions DCIDs of the form 'bio/{protein id}_{protein id}' (e.g. 'bio/P53_HUMAN_P53_HUMAN')
+    Duplicates include both exact duplicates and reverse duplicates.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' is duplicate with both itself and 'bio/CBP_HUMAN_P53_HUMAN'.
+    '''
+    dcids = set()
+    uniques = []
+    for interaction_dcid in interaction_dcids:
+        source_id, target_id = interactors(interaction_dcid)
+        # filter out self-interactions
+        if source_id == target_id:
+            continue
+        reverse_dcid = reverse_interaction_dcid(interaction_dcid)
+        if interaction_dcid in dcids or reverse_dcid in dcids:
+            uniques.append(interaction_dcid)
+        else:
+            dcids.update([interaction_dcid, reverse_dcid])
+    return uniques
+
+
+def symmetrized_scores(scores):
+    '''
+    Given a dict mapping interaction IDs to scores, return a new dict D of the same type such that
+    1) For any interaction ID in D, its reversal is also a key of D
+    2) Any interaction ID in D maps to the same value as its reversal
+    3) If an interaction ID is in scores but its reversal is not, D[interaction ID] = scores[interaction ID]
+    4) If both an interaction ID F and its reversal R are in scores, D[F] = D[R] = max(scores[F], scores[R])
+    '''
+    scores_sym = {}
+    for interaction_id, score in scores.items():
+        reverse_id = reverse_interaction_id(interaction_id)
+        reverse_score = scores.get(reverse_id, DEFAULT_INTERACTION_SCORE)
+        scores_sym[interaction_id] = scores_sym[reverse_id] = max(
+            score, reverse_score)
+    return scores_sym
+
+
+def flatten(nested_list):
+    '''
+    Given a list of lists, return a depth-1 flattened version
+    '''
+    flat = []
+    for L in nested_list:
+        flat.extend(L)
+    return flat
+
+
+def partition_expansion_cross(target_ids, node_set):
+    '''
+    Return two lists expansion_targets, cross_targets such that
+    1) expansion_targets U cross_targets = target_ids
+    2) expansion_targets and cross_targets are disjoint
+    3) all elements of cross_targets are in node_set
+    4) (implied) no elements of expansion_targets are in node_set
+    5) Stability: If id1 appears before id2 in target_ids and they are both assigned to the same
+       partition (either expansion_targets or cross_targets), id1 appears before id2 in this partition too.
+    '''
+    expansion_targets = []
+    cross_targets = []
+    for target_id in target_ids:
+        if target_id in node_set:
+            cross_targets.append(target_id)
+        else:
+            expansion_targets.append(target_id)
+    return expansion_targets, cross_targets
```

Comment:
Same comment about returning tuples as above. You could also return a map/dict which has the keys corresponding to each of these targets.

Reply:
Got it will do!